src/additional_zones.rs

```diff
@@ -0,0 +1,197 @@
+                    min: Coordinate {
+                        x: p.0.x - std::f64::EPSILON,
+                        y: p.0.y - std::f64::EPSILON,
+                    },
+                    max: Coordinate {
+                        x: p.0.x + std::f64::EPSILON,
+                        y: p.0.y + std::f64::EPSILON,
+                    },
+                });
+                zone.is_generated = true;
+                zones.push(zone);
+            }
+        }
+    }
+    zones
+}
+
+fn convert_to_geos(geom: GGeom) -> Option<MultiPolygon<f64>> {
+    match geom.try_into().expect("conversion to geo failed") {
+        geo::Geometry::Polygon(x) => Some(MultiPolygon(vec![x])),
+        y => {
+            if let Some(x) = y.into_multi_polygon() {
+                Some(x)
+            } else {
+                None
+            }
+        }
+    }
+}
+
+fn remove_from_existing_towns(zone: &mut Zone, towns: &[&Zone]) {
+    if towns.is_empty() {
+        return;
+    }
+    if let Some(ref mut boundary) = zone.boundary {
+        let mut updates = 0;
+        let mut g_boundary = boundary.try_into().expect("failed to convert to geos");
+        for town in towns {
+            if let Some(ref t_boundary) = town.boundary {
+                let g_t_boundary = t_boundary.try_into().expect("failed to convert to geos");
+                if g_boundary.intersects(&g_t_boundary).unwrap_or_else(|_| false) {
+                    if let Some(b) = g_boundary.difference(&g_t_boundary).ok() {
+                        updates += 1;
+                        g_boundary = b;
+                    }
+                }
+            }
+        }
+        if updates > 0 {
+            if let Some(g) = convert_to_geos(g_boundary) {
+                *boundary = g;
+            }
+        }
+    }
+}
+
+fn compute_voronoi(parent: &ZoneIndex, places: &[&Zone], zones: &[Zone], towns: &[&Zone]) -> Vec<Zone> {
+    let points: Vec<Point<_>> = places.iter()
+                                      .filter_map(|p| p.center)
+                                      .collect();
+    if points.len() == 1 {
+        let mut place = places[0].clone();
+        let parent = &zones[parent.index];
+
+        if parent.zone_type == Some(ZoneType::Country) {
+            place.boundary = Some(convert_to_geos(
+                                    place.center.as_ref()
+                                              .map(|x| x.try_into()
+                                                        .expect("failed to convert point"))
+                                              .unwrap()
+                                              .buffer(0.01, 2)
+                                              .expect("Failed to create a buffer"))
+                                  .expect("failed to convert to geo"));
+        } else {
+            place.boundary = parent.boundary.clone();
+        }
+        remove_from_existing_towns(&mut place, towns);
```

Comment:
minor: I'm not a fan of the name.

maybe `extrude_existing_town` ?

Reply:
I approve this new name! :)